packages/txservice/src/monitor/monitor.ts

```diff
@@ -0,0 +1,242 @@
+    this.chainId = this.provider.chainId;
+    // A separate loop will make sure they get through or get backfilled.
+    this.backfillLoop();
+  }
+
+  /**
+   * This will create a transaction with an assigned nonce as well as estimated gas / set gas price.
+   * We use this structure to essentially enforce all created transactions are saved locally in the buffer for
+   * continue monitoring, thus enabling us to further ensure they all go through.
+   * 
+   * @param minTx - Minimum transaction params needed to form a transaction for sending.
+   *
+   * @returns Transaction instance with populated params, ready for submit.
+   */
+  public async createTransaction(minTx: WriteTransaction): Promise<Transaction> {
+    // Estimate gas here will throw if the transaction is going to revert on-chain for "legit" reasons. This means
+    // that, if we get past this method, we can safely assume that the transaction will go through on submit, saving for
+    // instances where the provider malfunctions.
+    const gas = await this.getGas(minTx);
+    // Queue up the transaction with these values.
+    const result = await this.queue.add(async (): Promise<{ value: Transaction | Error; success: boolean }> => {
+      try {
+        // NOTE: This call must be here, serialized within the queue, as it is depenedent on current pending transaction count.
+        const nonce = await this.getNonce();
+        // Create a new transaction instance to track lifecycle.
+        const transaction = new Transaction(this.logger, this.provider, minTx, nonce, gas);
+        this.buffer.insert(nonce, transaction);
+        return { value: transaction, success: true };
+      } catch (e) {
+        // validate that the transaction failed on chain?
+        // didn't fail on chain? log this occurrence w/ error but ignore error ?!? -> WTF?
+        // if it did fail on chain, THROW
+        return { value: e, success: false };
+      }
+    });
+    if (result.success) {
+      return result.value as Transaction;
+    } else {
+      throw result.value;
+    }
+  }
+
+  private async getGas(transaction: WriteTransaction): Promise<Gas> {
+    const method = this.getGas.name;
+    // Get gas estimate.
+    let gasLimit: BigNumber;
+    let result = await this.provider.estimateGas(transaction);
+    if (result.isErr()) {
+      if (result.error instanceof TransactionReverted) {
+        // If we get a TransactionReverted error, that means the gas estimate call
+        // indicated our transaction would fail on-chain. The details of the failure will
+        // be included in the error.
+        throw result.error;
+      }
+      this.logger.warn(
+        {
+          method,
+          transaction: transaction,
+          error: result.error,
+        },
+        "Estimate gas failed due to an unexpected error.",
+      );
+      throw result.error;
+    } else {
+      gasLimit = result.value;
```

Comment:
(i know i already mentioned this but i resolved that comment so adding it here)

we should make sure we add some buffer to the gas limit here. like maybe 20% of the estimate or something. cc @rhlsthrm for best practice